modules/quest/src/main/java/unnamed/mmo/quest/objective/ParallelObjective.java

```diff
@@ -0,0 +1,79 @@
+package unnamed.mmo.quest.objective;
+
+import com.google.auto.service.AutoService;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import net.kyori.adventure.text.Component;
+import net.minestom.server.utils.NamespaceID;
+import net.minestom.server.utils.validate.Check;
+import org.jetbrains.annotations.NotNull;
+import unnamed.mmo.quest.QuestContext;
+import unnamed.mmo.quest.event.QuestObjectiveCompleteEvent;
+import unnamed.mmo.util.EventUtil;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static unnamed.mmo.util.ExtraCodecs.lazy;
+
+public record ParallelObjective(List<Objective> children) implements Objective {
+    public static final Codec<ParallelObjective> CODEC = RecordCodecBuilder.create(i -> i.group(
+            lazy(() -> Objective.CODEC).listOf().fieldOf("children").forGetter(ParallelObjective::children)
+    ).apply(i, ParallelObjective::new));
+
+    public ParallelObjective {
+        //todo not sure this should be an error? Probably should just complete instantly
+        Check.argCondition(children.isEmpty(), "children must not be empty");
+        Check.argCondition(children.size() > 32, "cannot have more than 32 children");
+    }
+
+    private static final Codec<Integer> COMPLETED = Codec.INT.orElse(0);
+
+    @Override
+    public @NotNull CompletableFuture<Void> onStart(@NotNull QuestContext context) {
+        CompletableFuture<?>[] futures = new CompletableFuture<?>[children.size()];
+
+        // Completed is a bitset of the completed children. A one at the bit index
+        // of a child indicates that the child has already been completed.
+        int completed = context.get(COMPLETED);
+        for (int i = 0; i < children.size(); i++) {
+            // If already complete, ignore it
+            if ((completed & (1 << i)) != 0) {
+                futures[i] = CompletableFuture.completedFuture(null);
+                continue;
+            }
+
+            // Not complete, restart this child.
+            Objective child = children().get(i);
+            QuestContext childContext = context.child(String.valueOf(i));
```

Comment:
A child could be in progress, it seems this method will restart it from scratch, or am I misreading this?

Reply:
Nope (there is a test for this: `TestParallelObjective#testSequenceResume`), this is the "genius" of quest context. It transparently handles de/serialization of child contexts, so `onStart` can act the same whether it is a new quest or an old one. In this case, the context has a child entry for `i` with its last serialized state, so when `onStart` is called, it has its last data.